design-document/design-document.md

```diff
@@ -0,0 +1,131 @@
+ 
+ - serves web assets
+ - manages user sessions
+ - returns contents of a directory on a given path
+
+Finally, a few constraints and allowed/encouraged shortucts:
+
+ - containerize the app (Docker)
+ - use native browser API for networking
+ - use native/standard libraries
+ - store users & sessions in memory, no need for a db
+ - feel free to hardcode usernames, hashes, or directory paths
+ - make it secure (strong TLS, no web security vulnerabilities)
+
+# Wireframes
+
+_Note: Honestly, these are more colourful and specific than wireframes need to be — or maybe even should be — but it was too fast & easy with [Whimsical](https://whimsical.com) to do it any other way._
+
+<img src="assets/signin.png" style="max-width: 600px" alt="Sign In" />
+<img src="assets/browse.png" style="max-width: 600px;" alt="Browse a Directory" /> 
+
+# Implementation Details
+
+The entire codebase shall use typescript, reasonable linting, and be dockerized.
+
+In a real app, error reporting would be done through an integration (like Sentry), but for the purposes of the challenge `console.error` will suffice. 
+
+There will be a few tests for the key React components and for the key API endpoints, but coverage shouldn't become an obsession.
+
+## Frontend
+
+Let's use React (`create-react-app` for the simplicity of its setup), React Router, and Tailwind on the frontend. Core considerations:
+
+ - **Authentication:** All protected routes will be wrapped in a simple component `<Protected />` which checks whether the user is authenticated. When they're not, the user is taken to the `/sign-in` page, and the original URL is stored in the `state` property on the `location` object of our router. That way, we can redirect the user back to where they intended to go once they've signed in.
+ - **URL as the source of truth:** We will store user's current working directory inside the file tree as URL params, the sorting and filtering criteria as query params (`/browse/my-directory/my-child-directory?sort=name-up&search=holida`).
+ - **Error handling**: Let's pay attention to loading states, errors (at the very least 400, 401, 403, 404, 500) and edge cases such as invalid filtering options, etc.
+ - **Security:** Input and output shall be sanitised (URL params, query params, and the search input field).
+
+## Backend
+
+The server will be a plain `express` app. For the sake of simplicity, let's skip a database and store everything in memory instead.
+
+Since searching and filtering is required on the client only, a handful of API endpoints will suffice:
+
+ - `GET /contents?path={path}` to obtain 1-level deep contents of a directory, returning `{ contents: Item }` where `interface Item { name: string, sizeKb: number | undefined, type: "file" | "dir", items: undefined | Item[] }`
+ - `POST /sign-in` to sign the user in
+ - `POST /sign-out` to sign the user out (`POST` to avoid [prefetching it](https://twitter.com/nick_craver/status/296281730984316928?lang=fa) 🤦) 
+
+## Security
```

Comment:
A few more questions:

* What password hashing algorithm will you use?
* How will you handle open redirects?

Reply:
Ah, right, sorry for not going more into detail on these:

1. I will use `crypto.pbkdf2(password, salt, 10000, 64, 'sha512')` to compute the password hash, saving the `salt` alongside the result as a user property. The salt needs to be there in order to prevent identical hashes of identical passwords, the work factor / iteration count to make it even more difficult for an attacker to compute hashes of all possible passwords.
2. Hmm, is there an open redirect vulnerability? The only place I can think where a redirect is happening is when signing in, but a) it's state-based and, more importantly, b) the destination comes from the URL the user visited on **our** website. Am I missing something obvious here?

Reply:
1. PBKDF2 is fine, but I am curious why you picked it.

2. Right, but what if the attacker tricks you into clicking on `https://example.com?redirect=evil.com/malware`?

Reply:
1. In my understanding, `pbkdf2` is much slower than, for example, computing an HMAC hash with, say, `sha256`. This is by design to make pre-computing passwords difficult, making brute-force attacks harder. Now that I checked the `crypto` module properly, there's also `scrypt` which internally uses `pbkdf2` **and** is designed to require a lot more memory, making password pre-computation difficult even in parallel.
2. That's what I mean, the redirect won't appear in the URL at all, it will be purely state-based: come to a page -> not logged in -> save the state internally in the router -> redirect to the log in page -> log in -> redirect to the original URL based on the internal state. In other cases, I think URL is a wonderful source of truth, but here I figured this is a better approach for this exact reason, removing the possibility of the vulnerability altogether. If we were to protect against open-redirects, I would probably go for the domain-based approach: on the server, just `'https://our-domain.com' + redirectUrl`, on the client, use `history.push(redirectUrl)` which will never go to another origin.